src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java

```diff
@@ -340,4 +352,60 @@ private static void assertStepExecutions(FlowExecution e, String... steps) throw
         }
     }
 
+    @Test public void trustedShell() {
+        trustedShell(true);
+    }
+
+    @Test public void trustedShell_control() {
+        trustedShell(false);
+    }
+
+    /**
+     * Insert trusted/ dir into the trusted shell to enable trusted code execution
+     */
+    @TestExtension("trustedShell")
+    public static class TrustedShell extends GroovyShellDecorator {
+        @Override
+        public GroovyShellDecorator forTrusted() {
+            return new UntrustedShellDecorator();
+        }
+    }
+
+    @TestExtension("trustedShell_control")
+    public static class UntrustedShellDecorator extends GroovyShellDecorator {
+        @Override
+        public void configureShell(@CheckForNull CpsFlowExecution context, GroovyShell shell) {
+            try {
+                URL u = TrustedShell.class.getClassLoader().getResource("trusted/foo.groovy");
+                shell.getClassLoader().addURL(new URL(u,"."));
+            } catch (MalformedURLException e) {
+                throw new AssertionError(e);
+            }
+        }
+    }
+
+    private void trustedShell(final boolean pos) {
+        SECRET = false;
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition("new foo().attempt()", true));
+                WorkflowRun b = story.j.waitForCompletion(p.scheduleBuild2(0).get());
+                if (pos) {
+                    story.j.assertBuildStatusSuccess(b);
+                    assertTrue(SECRET);
+                } else {
+                    // should have failed with RejectedAccessException trying to touch 'SECRET'
+                    story.j.assertBuildStatus(Result.FAILURE, b);
+                    story.j.assertLogContains("RejectedAccessException: Scripts not permitted to use staticField org.jenkinsci.plugins.workflow.cps.CpsFlowExecutionTest SECRET",b);
```

Comment:
:ant: Can do various things to avoid hardcoding this much text here: use `CpsFlowExecutionTest.class.getName()`; create an actual `RejectedAccessException` and `toString` it (IIRC)